Sync script updates

Bankso opened the pull request:
Updating sync_publications.py to allow CSVs to be synced to CCKP database

- support use of staging CSVs, instead of tables
- extract 'theme', 'grantName', and 'consortium' information from Grants - Merged table and add to each publication listed
- default manifest set as "syn53478776", the current synId for the PublicationView staging database CSV
- remove dataset alias column conversion to string, since reading from CSV appears to produce strings

Bankso committed:
Update sync_publications.py

Script modifications to:
- support use of staging CSVs, instead of tables
- extract 'theme', 'grantName', and 'consortium' information from Grants - Merged table and add to each publication listed
- default manifest set as "syn53478776", the current synId for the PublicationView staging database CSV

Bankso committed:
Update sync_publications.py

- remove dataset alias column conversion to string, since reading from CSV appears to produce strings

vpchung commented on portal_tables/sync_publications.py:
```diff
@@ -34,81 +37,109 @@ def add_missing_info(pubs, grants):
+        for row in pubs.itertuples():
+            i = row[0]
+            n = row[4].split(",")
+            extracted = []
```
On the rare occasion there are duplicate themes/consortium between the grants, could we maybe use a set here instead of a list?

Then, below, we can add items to the set with `s.update(new_item)` instead of `l.append(new_item)`.

Bankso replied:
Thank you for bringing this up! The duplicate themes and consortium names between grants is definitely an issue I've seen. I added [line 67](https://github.com/mc2-center/mc2-center-dcc/pull/40/files#diff-3f2f07b16780917689682ca0af91d5420dae7335439d92ae28e201f319cd3532R67) to address the issue of duplicate entries:

`clean_values = ', '.join(list((dict.fromkeys(extracted))))`

Do you think this will hold up, or is the set method you described better suited here?

vpchung replied:
Oh!  I actually didn't explore that line of code, so let me try real quick.  If it works, then let's ignore my set suggestion.

vpchung reviewed (approved):
Overall, looks good to me!  I have one minor suggestion, but it may not be relevant.

vpchung commented on portal_tables/sync_publications.py:
```diff
@@ -34,81 +37,109 @@ def add_missing_info(pubs, grants):
+                    print(f"No match found for grant number: {g}")
+                    continue
+            
+            clean_values = ', '.join(list((dict.fromkeys(extracted))))
```
```suggestion
            clean_values = ', '.join(list(dict.fromkeys(extracted)))
```

vpchung commented on portal_tables/sync_publications.py:
```diff
@@ -34,81 +37,109 @@ def add_missing_info(pubs, grants):
+                    print(f"No match found for grant number: {g}")
+                    continue
+            
+            clean_values = ', '.join(list((dict.fromkeys(extracted))))
```
We may need to add an additional check here before joining the list as a string.  IIRC, `theme` and `consortium` are STRINGLIST in the Merged table, so we would want to save `clean_values` as a list rather than a string.

vpchung replied:
Oh!  Perhaps a check is not needed after all, as `grantName` is also a STRINGLIST.  In that case, we should remove the `join()`

vpchung commented on portal_tables/sync_publications.py:
```diff
@@ -34,81 +37,109 @@ def add_missing_info(pubs, grants):
-            new_pubs = add_missing_info(new_pubs.copy(), grants)
-            sync_table(syn, new_pubs.copy(), args.portal_table)
-    print("DONE ✓")
+    print("\n\nCSV downloaded from Synapse:\n", manifest)
```
Do you think it'd be valuable to add another flag (maybe something like `-v`, `--verbose`) and if passed, then output the `manifest`, `database`, and `new_table`?  Otherwise, only print:

```
$ python portal_tables/sync_publications.py
Inputs will be processed and provided for review 

Database will NOT be updated.

Processing publications staging database...

Copy of final table stored at: ./final_table.csv
DONE ✓
```

Bankso committed:
Remove join from clean_values definition

Populates new columns with lists, instead of strings

vpchung commented on portal_tables/sync_publications.py:
```diff
@@ -34,81 +37,109 @@ def add_missing_info(pubs, grants):
-        .str.join(", ")
-    )
+    if dryrun:
+        print(u"\u26A0", "WARNING:",
```
Perhaps remove this message, as it's already achieved with the newer message?  Otherwise, this is sandwiched between the tables:

![Screenshot 2024-02-16 at 1 49 57 PM](https://github.com/mc2-center/mc2-center-dcc/assets/9377970/d03c5f4f-fd72-41ee-a10f-74ac92c709d4)

Bankso committed:
Update sync_publications.py

- added argument '-v'/'--verbose' to toggle if tables are printed to command line at each processing step
- added 'if' statements to make use of '-v' input flag
- removed redundant dryrun flag warning

Bankso committed:
Update sync_publications.py

Remove extra ()

vpchung commented on portal_tables/sync_publications.py:
```diff
@@ -87,16 +91,16 @@ def sync_table(syn, pubs, table, dryrun):
         table_rows = pubs.values.tolist()
         syn.store(Table(schema, table_rows))
 
+    else:
```
nit: we don't need an `else` since no coding logic is provided here.

Bankso committed:
Update sync_publications.py

Remove else statement from 'sync_table' function

vpchung commented on portal_tables/sync_publications.py:
```diff
@@ -34,81 +41,113 @@ def add_missing_info(pubs, grants):
-
-    new_rows = pubs.values.tolist()
-    syn.store(Table(schema, new_rows))
+    if dryrun is False:
```
nit: it's generally considered bad practice to compare directly with `True`/`False`. Instead, use:

`if variable:`

or 

`if not variable:`

to compare truthy/falsy values.  Doing this also executes faster!

Bankso replied:
Good call! I had `if not dryrun` at first and then changed it without thinking, so thank you for noting this!

Bankso committed:
Update sync_publications.py

Change dryrun check in sync_tables to use 'if not' notation instead of 'True/False' comparison

vpchung reviewed (approved):
LGTM! 🚀
